[TLI] Add mappings to SLEEF/ArmPL libcall variants taking linear args.

labrinea opened the pull request:
The mappings correspond to vectorized variants (fixed/scalable) for the math functions: modf, sincos, sincospi.

mgabka commented:
could you also add tests for the replace-with-veclib-pass which is using those mapping to:
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef-scalable.ll
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef.ll
llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-armpl.ll and

mgabka commented on llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:
```diff
@@ -0,0 +1,275 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call.*(modf|modff|sincos|sincosf|sincospi|sincospif)" --version 4
```
I was rather aiming to keep the tests in :
llvm/test/Transforms/LoopVectorize/AArch64/sleef-calls-aarch64.ll
llvm/test/Transforms/LoopVectorize/AArch64/armpl-calls.ll

so we have a single file per library, what do you think?

labrinea replied:
The mentioned test files have been refactored in #76146.

mgabka commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -504,7 +504,8 @@ void VPWidenCallRecipe::execute(VPTransformState &State) {
     SmallVector<Type *, 2> TysForDecl;
     // Add return type if intrinsic is overloaded on it.
-    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1)) {
+    if (isVectorIntrinsicWithOverloadTypeAtArg(VectorIntrinsicID, -1) &&
```
this is a LV bug, so I think it deserves a separate commit with target independent test, probably extension to the tests added in https://github.com/llvm/llvm-project/pull/73941 as all of the tests there were using functions  returning a value.

labrinea replied:
This fix has been moved to a separate PR as suggested: #76274

labrinea commented:
> could you also add tests for the replace-with-veclib-pass which is using those mapping to: llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef-scalable.ll llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-sleef.ll llvm/test/CodeGen/AArch64/replace-intrinsics-with-veclib-armpl.ll

I am not seeing corresponding llvm instrinsics for the veclib functions I've added. I don't think there's anything needed to be done in theses test files.

mgabka commented on llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:
```diff
@@ -0,0 +1,275 @@
+  %num = load double, ptr %gepa, align 8
+  %gepb = getelementptr double, ptr %b, i64 %indvars.iv
+  %data = call double @modf(double %num, ptr %gepb)
+  %gepc = getelementptr inbounds double, ptr %c, i64 %indvars.iv
+  store double %data, ptr %gepc, align 8
```
this isn't needed

labrinea replied:
Without a store we are not using the return value of the call. I see no harm here.

mgabka replied:
sure, I just pointed it out as it could simplify the test.

mgabka commented on llvm/test/Transforms/LoopVectorize/AArch64/vector-libcall-linear-args.ll:
```diff
@@ -0,0 +1,275 @@
+define void @test_modf(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %n) {
+; SLEEF-NEON-LABEL: define void @test_modf(
+; SLEEF-NEON-SAME: ptr noalias [[A:%.*]], ptr noalias [[B:%.*]], ptr noalias [[C:%.*]], i64 [[N:%.*]]) #[[ATTR0:[0-9]+]] {
+; SLEEF-NEON:    [[TMP8:%.*]] = call <2 x double> @_ZGVnN2vl8_modf(<2 x double> [[WIDE_LOAD:%.*]], <2 x ptr> [[TMP6:%.*]])
```
these tests are just showing that there is a compiler bug.

In this particular case you can see that it is passing <2x ptr> to the function instead of just ptr, the whole idea of linear pointers is that you are passing just a ptr to the function and the function will increment it inside according to the step.

I am not sure where is the bug, the tests added by @huntergr-arm  were using only integer types from what I can see https://github.com/llvm/llvm-project/pull/73941/files#diff-9e8f66d7297707a074a06db9f068f2bd0cff25d12e4fc07def062d224416aff4

Could you please investigate it?

labrinea replied:
Fixed in https://github.com/llvm/llvm-project/pull/76753

labrinea committed:
[TLI] Add mappings to SLEEF/ArmPL libcall variants taking linear args.

The mappings correspond to vectorized variants (fixed/scalable) for the
math functions: modf, sincos, sincospi.
